File: src/mvmctl/models/image.py
"""Image data models."""

from __future__ import annotations

import platform
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from mvmctl.utils.common import CommonUtils

if TYPE_CHECKING:
    from mvmctl.models.vm import VMInstanceItem


@dataclass
class ImageItem:
    """Image record — maps to images table."""

    id: str
    type: str
    name: str
    arch: str
    path: str
    fs_type: str
    minimum_rootfs_size_mib: int
    original_size: int
    is_default: bool
    is_present: bool
    pulled_at: str
    created_at: str
    updated_at: str

    distro: str | None = None
    fs_uuid: str | None = None
    compressed_size: int | None = None
    compression_ratio: float | None = None
    compressed_format: str | None = None
    deleted_at: str | None = None
    vms: list[VMInstanceItem] | None = None

    def __post_init__(self) -> None:
        """Coerce bool fields loaded from SQLite."""
        CommonUtils.coerce_bool_fields(self, {"is_default", "is_present"})


@dataclass
class ImageSpec:
    id: str
    image_type: str
    version: str
    name: str
    source: str
    format: str  # noqa: N816
    arch: str = field(default_factory=platform.machine)
    sha256: str | None = None
    sha256_url: str | None = None
    list_url_template: str | None = None
    size: int | None = None


@dataclass
class ImageVersion:
    """A published version of an image type from an upstream provider.

    NOT frozen — mutable for convenience. Returned by version resolvers
    to describe an available download for a given image type.
    """

    version: str
    codename: str | None
    type: str
    download_url: str
    sha256_url: str | None
    format: str
    display_name: str = ""
    type_name: str = ""

File: internal/lib/model/image.go
package model

// ── ImageItem ──

// ImageItem corresponds to Python's ImageItem dataclass exactly.
type ImageItem struct {
	ID               string   `json:"id"                          db:"id"`
	Type             string   `json:"type"                        db:"type"`
	Name             string   `json:"name"                        db:"name"`
	Arch             string   `json:"arch"                        db:"arch"`
	Path             string   `json:"path"                        db:"path"`
	FSType           string   `json:"fs_type"                     db:"fs_type"`
	MinRootfsSizeMiB int      `json:"minimum_rootfs_size_mib"     db:"minimum_rootfs_size_mib"`
	OriginalSize     int64    `json:"original_size"               db:"original_size"`
	IsDefault        bool     `json:"is_default"                  db:"is_default"`
	IsPresent        bool     `json:"is_present"                  db:"is_present"`
	PulledAt         string   `json:"pulled_at"                   db:"pulled_at"`
	CreatedAt        string   `json:"created_at"                  db:"created_at"`
	UpdatedAt        string   `json:"updated_at"                  db:"updated_at"`
	Version          string   `json:"version"                     db:"version"`
	Distro           string   `json:"distro,omitempty"            db:"distro"`
	FSUUID           string   `json:"fs_uuid,omitempty"           db:"fs_uuid"`
	CompressedSize   *int64   `json:"compressed_size,omitempty"   db:"compressed_size"`
	CompressionRatio *float64 `json:"compression_ratio,omitempty" db:"compression_ratio"`
	CompressedFormat *string  `json:"compressed_format,omitempty" db:"compressed_format"`
	DeletedAt        *string  `json:"deleted_at,omitempty"        db:"deleted_at"`

	// Resolved relations
	VMs []*VM `json:"vms,omitempty"`
}

// ── ImageSpec ──

// ImageSpec corresponds to Python's ImageSpec dataclass.
type ImageSpec struct {
	Type            string  `yaml:"type"`
	Version         string  `yaml:"version"`
	Name            string  `yaml:"name"`
	Source          string  `yaml:"source"`
	Format          string  `yaml:"format"`
	Arch            string  `yaml:"arch"`
	SHA256          string  `yaml:"sha256,omitempty"`
	SHA256URL       string  `yaml:"sha256_url,omitempty"`
	ListURLTemplate *string `yaml:"list_url_template,omitempty"`
	Size            *int64  `yaml:"size,omitempty"`
}
